File: src/ruby/hackerrank-ruby-tutorial/closure_v3.rb
def add1(f)
  f.call
end

x = 1

fn = -> { x + 1 }

x = 10

p add1(fn)

#
# Which `x` will the closure remember?
#
# x = 10 will be remembered. The output is 11.
#
# So, if `x` is defined earlier, and then reassigned, it remembers
# its last value. But as we saw in the previous example, if it is
# only defined after we declare the closure, then it throws an error.
#
# Tricky stuff to remember and avoid.
#

File: src/ruby/hackerrank-ruby-tutorial/closure_v3.js
const log = console.log.bind(console);

var x = 1;

function f() {
  return x + 1
}

var x = 10;

log(f());

//
// Which `x` will be remembered by the closure created by `f`?
//
// It will be the second one. The output is 11.
//
// Because of hoisting, this is different than closure_v3.rb
// which would cause an error here saying `x` is not defined.
//
